Board editor: Improve the "add via" tool

5n8ke opened the pull request:
General UX and Code improvements for the "add via" command

# General improvements
The via is now combined with NetPoints/NetLines of its specified signal. Adding a via at the position of a pad, other via or an intersection of multiple different (!) signals is no longer possible.

# Auto Mode
In "Auto" Mode, the via is added to an existing NetSignal at the target location. Otherwise, it defaults to the signal with the most segments but should be changed to the closest signal.

# CmdSplitNetline
A new UndoCommandGroup, that splits the NetLine at the given position. The position does not have to be **on** the netline.

ubruhin commented:
Nice! Some comments so far:

- If the schematic is empty (and thus no net signals exist), the application crashes when starting the "add via" tool. Previously we rejected starting the tool if no net signals exist to avoid this issue.
- When the net signal is set to "auto", performance is heavily reduced (at least on a larger board with many elements). It seems that the calculations in the MouseMoveEvent are quite heavy :( It might be worth considering a more efficient way to determine the closest netsignal. However, I didn't test it with a release build yet, but only with a debug build.
- I'd prefer to rename the `CmdSplitNetLine` class to something containing the term "Board" to make clear it only splits *board* netlines.
- Removing `mUndoCmdActive` and instead calling `mUndoStack.isCommandGroupActive()` is problematic since now you don't know whether the active undo command group was started by the "add via" tool, or by some other tool. It could lead to serious issues if another tool started the undo command group, but the "add via" tool "thinks" it was started by itself.
- If there exists a net with name "Auto", choosing that signal in the dropdown menu doesn't enforce the corresponding net because the "auto net" feature only matches the text in the dropdown menu :wink:

5n8ke commented:
@ubruhin 
Thanks for your feedback. I addressed almost all of your comments.
The low performance of the auto command is not yet considered. I think for now the addition of the auto mode is useful and an improved signal search algorithm would be part of a new project.

ubruhin commented on libs/librepcb/projecteditor/boardeditor/fsm/bes_addvia.cpp:
```diff
@@ -283,45 +292,65 @@ BES_Base::ProcRetVal BES_AddVia::processSceneEvent(BEE_Base* event) noexcept {
-  Q_ASSERT(mUndoCmdActive == false);
-  Q_ASSERT(mCurrentViaNetSignal);
+  Q_ASSERT(mSubState == SubState_Idle);
+  Q_ASSERT(!mUndoStack.isCommandGroupActive());
```
This assertion fails (and thus LibrePCB crashes) if another tool (for example in the schematic editor) is active when the "add via" tool is started. It's totally legal to try starting a tool while another command is active, thus this assertion should be removed:

```suggestion
```

ubruhin commented on libs/librepcb/projecteditor/boardeditor/fsm/bes_addvia.cpp:
```diff
@@ -283,45 +292,65 @@ BES_Base::ProcRetVal BES_AddVia::processSceneEvent(BEE_Base* event) noexcept {
+  Q_ASSERT(!mUndoStack.isCommandGroupActive());
 
   try {
+    mSubState = SubState_PositioningVia;
     mUndoStack.beginCmdGroup(tr("Add via to board"));
```
The state must be set *after* starting the new undo command, since starting a command raises an exception if there is already a command active. Then we don't want to be in `SubState_PositioningVia` state since this would make `abortPlacement()` aborting a command which was started by another tool.

```suggestion
    mUndoStack.beginCmdGroup(tr("Add via to board"));
    mSubState = SubState_PositioningVia;
```

5n8ke replied:
Will fix it there too

ubruhin commented on libs/librepcb/projecteditor/boardeditor/fsm/bes_addvia.cpp:
```diff
@@ -283,45 +292,65 @@ BES_Base::ProcRetVal BES_AddVia::processSceneEvent(BEE_Base* event) noexcept {
+      }
+    }
+  }
+  Q_ASSERT(!mUndoStack.isCommandGroupActive());
```
Same here - if we are not in `SubState_PositioningVia`, we must not rely on whether no command is active currently.

```suggestion
```

ubruhin commented on libs/librepcb/projecteditor/boardeditor/fsm/bes_addvia.cpp:
```diff
@@ -330,22 +359,115 @@ bool BES_AddVia::updateVia(Board& board, const Point& pos) noexcept {
   try {
     mViaEditCmd->setPosition(pos, false);
+
+    Board* board = &mCurrentVia->getBoard();
```
That looks strange - if you get a reference to a board, you should store it as a reference since this way it's clear that it can never be null.

```suggestion
    Board& board = mCurrentVia->getBoard();
```

But actually I would even pass `Board& board` as a parameter to the `fixVia()` method (just like `updateVia()`) instead of fetching it from the currently placed via. The only function which calls `fixVia()` even already has a pointer to the board.

5n8ke replied:
Done.

ubruhin commented on libs/librepcb/projecteditor/cmd/cmdboardsplitnetline.cpp:
```diff
@@ -0,0 +1,81 @@
+namespace editor {
+
+CmdBoardSplitNetLine::CmdBoardSplitNetLine(BI_NetLine& netline, Point& pos)
+  noexcept : UndoCommandGroup(tr("Split netline")),
```
For translated strings, we don't use the term "netline" since translators and users might not know what it means. Instead, for board netlines we use the term "trace".

```suggestion
  noexcept : UndoCommandGroup(tr("Split trace")),
```

ubruhin commented on libs/librepcb/projecteditor/cmd/cmdboardsplitnetline.cpp:
```diff
@@ -0,0 +1,81 @@
+  execNewChildCmd(cmdAdd.take()); // can throw
+  execNewChildCmd(cmdRemove.take()); // can throw
+
+  undoScopeGuard.dismiss();  // no undo required
```
FYI, since this undo command group does not contain logic which depends on the executed commands, it would be simpler (and thus possibly safer) to append the commands with `appendChild()` instead of `execNewChildCmd()` and then calling `UndoCommandGroup::performExecute()` at the end of the function. This way you don't need the undo scope guard at all. See for example `CmdRemoveSelectedSymbolItems`.

But it should work this way, so no need to change it now...

5n8ke replied:
Thanks for the info. Since I'm still working on it, I will change it.

ubruhin commented on libs/librepcb/projecteditor/cmd/cmdboardsplitnetline.h:
```diff
@@ -0,0 +1,76 @@
+class CmdBoardSplitNetLine : public UndoCommandGroup {
+public:
+  // Constructors / Destructor
+  explicit CmdBoardSplitNetLine(BI_NetLine& netline, Point& pos) noexcept;
```
The `Point` should be passed as a const reference:

```suggestion
  explicit CmdBoardSplitNetLine(BI_NetLine& netline, const Point& pos) noexcept;
```

ubruhin commented on libs/librepcb/projecteditor/cmd/cmdboardsplitnetline.h:
```diff
@@ -0,0 +1,76 @@
+
+  // Private Member Variables
+  BI_NetLine& mOldNetLine;
+  Point& mSplitPosition;
```
This looks quite dangerous. The `Point` object passed to the constructor will most probably have a shorter object lifetime than this undo command object, which will lead to dangling references and thus application crashes. It's much safer to store it by value. But since it is actually never accessed at all, let's remove it completely:

```suggestion
```

5n8ke replied:
I missed that...

ubruhin commented on libs/librepcb/projecteditor/cmd/cmdboardsplitnetline.h:
```diff
@@ -0,0 +1,76 @@
+namespace editor {
+
+/*******************************************************************************
+ *  Class CmdBoardNetLineSplit
```
Inconsistent naming:

```suggestion
 *  Class CmdBoardSplitNetLine
```

ubruhin commented on libs/librepcb/projecteditor/cmd/cmdboardsplitnetline.h:
```diff
@@ -0,0 +1,76 @@
+ *  Class CmdBoardNetLineSplit
+ ******************************************************************************/
+
+class CmdBoardSplitNetLine : public UndoCommandGroup {
```
For consistency, at least a very short doxygen comment should be added (I guess this would really be the first class with no doxygen comment at all):

```suggestion
/**
 * @brief Undo command to split a ::librepcb::project::BI_NetLine
 */
class CmdBoardSplitNetLine : public UndoCommandGroup {
```

ubruhin commented:
> The low performance of the auto command is not yet considered. I think for now the addition of the auto mode is useful and an improved signal search algorithm would be part of a new project.

Now I just tested it with a release build (on a decent PC) - unfortunately the board editor still lags on a medium complex board. I suspect on slower computers or with more complex boards, it will lag so much that it would be very annoying.

Therefore I'm really sorry, but I don't agree that the performance issue could be fixed at any time later. I'd really prefer to (at least try to) fix the performance issues right in this PR. But instead of optimizing the netsignal finder algorithm, for now we might simply run that algorithm less often. Example:

- When the "auto" mode is selected, start a `QTimer` with e.g. 100ms interval (to be determined by experiments)
- In the MouseMove event, do not immediately find the closest netsignal, but only set a flag to do so
- Within the timer callback, check that flag - if it is set, find the closest net signal, update the via and clear the flag.

I suspect this would at least avoid having a totally lagging UI, while keeping the implementation effort low. What do you think about that?

5n8ke commented on libs/librepcb/projecteditor/boardeditor/fsm/bes_addvia.cpp:
```diff
@@ -71,7 +79,7 @@ BES_AddVia::BES_AddVia(BoardEditor& editor, Ui::BoardEditor& editorUi,
 
 BES_AddVia::~BES_AddVia() {
-  Q_ASSERT(mUndoCmdActive == false);
+  Q_ASSERT(!mUndoStack.isCommandGroupActive());
```
Thanks. I fixed it.

5n8ke commented on libs/librepcb/projecteditor/boardeditor/fsm/bes_addvia.cpp:
```diff
@@ -91,19 +99,12 @@ bool BES_AddVia::entry(BEE_Base* event) noexcept {
   Q_UNUSED(event);
   Board* board = mEditor.getActiveBoard();
   if (!board) return false;
+  if (mEditor.getProject().getCircuit().getNetSignals().count() == 0)
```
Found also some other places with the same issue.

5n8ke commented:
Thanks for the many comments. I fixed them all.

I improved the finder a bit by not always calculating the most used NetSignal, but instead storing the information in mLastClosestNetSignal. I also implemented a timer, that allows the closestNetSignal to only be calculated every 500ms. It still causes a noticeable choppiness but only twice a second :/ ...
The finder algorithm only stores a new NetSignal, when it is directly found below the cursor, as previously.

ubruhin commented on libs/librepcb/projecteditor/boardeditor/fsm/bes_addvia.cpp:
```diff
@@ -368,31 +485,116 @@ void BES_AddVia::sizeEditValueChanged(const PositiveLength& value) noexcept {
-    mCurrentViaNetSignal = netsignal;
+    if (!netsignal) {
+      throw LogicError(__FILE__, __LINE__);
+    }
```
During testing, sometimes the via position was not updated at all when moving the cursor, which seemed to be very strange since `setPosition()` was called properly. After some debugging, it turned out that the `setNetSignal()` is now called every time in the MouseMove event, which constantly removes the whole net segment and re-adds it immediately, which seems to lead to graphical issues (in my case, it was just never repainted, although I'd expect flickering...). Updating the net signal only when needed, fixed that issue:

```suggestion
    }
    if (netsignal == &mCurrentVia->getNetSignalOfNetSegment()) {
      return;  // net signal is already correct, no update needed
    }
```

ubruhin commented on libs/librepcb/projecteditor/boardeditor/fsm/bes_addvia.cpp:
```diff
@@ -368,31 +485,116 @@ void BES_AddVia::sizeEditValueChanged(const PositiveLength& value) noexcept {
+      mLastClosestNetSignal = &atPosition->getNetSignalOfNetSegment();
+    }
+    mFindClosestNetSignal = false;
+    QTimer *timer = new QTimer(this);
+    connect(timer, &QTimer::timeout, [this]() {
+      this->mFindClosestNetSignal = true;
+    });
```
This regularly instantiates new `QTimer` objects which will be deleted only when closing the board editor. So the memory consumption constantly increases. I'd suggest to either re-use the same timer instance (by storing it as a member variable), or delete it after it expired:

```suggestion
    QTimer *timer = new QTimer(this);
    connect(timer, &QTimer::timeout, [this, timer]() {
      this->mFindClosestNetSignal = true;
      timer->deleteLater();
    });
```

ubruhin commented on libs/librepcb/projecteditor/boardeditor/fsm/bes_addvia.cpp:
```diff
@@ -330,22 +355,114 @@ bool BES_AddVia::updateVia(Board& board, const Point& pos) noexcept {
+    foreach (BI_NetLine* netline, board.getNetLinesAtScenePos(pos)) {
+      if (&netline->getNetSignalOfNetSegment() != netsignal) {
+        throw RuntimeError(__FILE__, __LINE__,
+          tr("Netline of a different signal already present"
+             "at target position."));
```
Missing space:

```suggestion
          tr("Netline of a different signal already present "
             "at target position."));
```

ubruhin commented on libs/librepcb/projecteditor/boardeditor/fsm/bes_addvia.cpp:
```diff
@@ -330,22 +355,114 @@ bool BES_AddVia::updateVia(Board& board, const Point& pos) noexcept {
+    foreach (BI_NetPoint* netpoint, board.getNetPointsAtScenePos(pos)) {
+      if (&netpoint->getNetSignalOfNetSegment() != netsignal) {
+        throw RuntimeError(__FILE__, __LINE__,
+          tr("Netpoint of a different signal already present"
+             "at target position."));
```
Missing space:

```suggestion
          tr("Netpoint of a different signal already present "
             "at target position."));
```
